// moontech-backend-library/libs/moontech-backend-library/src/alerts/alerts.module.ts
import { DynamicModule, Module } from '@nestjs/common';
import { PassportModule } from '@nestjs/passport';
import { AlertConfig } from './alert-config';
import { AlertsPatternController } from './alerts-pattern.controller';
import { AlertsPatternsService } from './alerts-patterns.service';
import { AlertsController } from './alerts.controller';
import { AlertsService } from './alerts.service';

@Module({
  imports:[
  ],
  controllers: [],
  providers: [],
  exports:[]
})
export class AlertsModule {
  static register(options?: AlertConfig): DynamicModule {
    
    const dynamicModule: DynamicModule = {
      module: AlertsModule,
    }
    if(options.endpoints && !options.patterns){
      dynamicModule.imports = [
        PassportModule.register({ defaultStrategy: 'jwt' })
      ];
      dynamicModule.controllers = [
        AlertsController
      ]; 
      dynamicModule.providers = [
        {
          provide: 'CONFIG_OPTIONS',
          useValue: options,
        },
        {
          provide: 'SERVICE',
          useClass: AlertsService,
          inject: [AlertsService]
        },
        AlertsService
      ];
      dynamicModule.exports = [
        AlertsService
      ];
    }else if (options.endpoints && options.patterns){
      dynamicModule.imports = [
        PassportModule.register({ defaultStrategy: 'jwt' })
      ];
      dynamicModule.controllers = [
        AlertsController
      ];
      dynamicModule.providers = [
        {
          provide: 'CONFIG_OPTIONS',
          useValue: options,
        },
        {
          provide: 'SERVICE',
          useClass: AlertsPatternsService,
          inject: [AlertsPatternsService]
        },
        AlertsPatternsService
      ];
      dynamicModule.exports = [
        AlertsPatternsService
      ];
    } else if(!options.endpoints && options.patterns) {
      dynamicModule.controllers = [
        AlertsPatternController
      ]; 
      dynamicModule.providers = [
        {
          provide: 'CONFIG_OPTIONS',
          useValue: options,
        },
        AlertsService,
      ];
      dynamicModule.exports = [
        AlertsService
      ];
    }

    return dynamicModule;
  }
}
