// moontech-backend-library/libs/moontech-backend-library/src/agvs/agvs.module.ts
import { DynamicModule, Module } from '@nestjs/common';
import { PassportModule } from '@nestjs/passport';
import { AgvConfig } from './agv-config';
import { AgvsPatternsController } from './agvs-patterns.controller';
import { AgvsPatternsService } from './agvs-patterns.service';
import { AgvsController } from './agvs.controller';
import { AgvsService } from './agvs.service';

@Module({
    imports: [],
    controllers: [],
    providers: [],
    exports: []
})
export class AgvsModule {
  static register(options?: AgvConfig): DynamicModule {
    
    const dynamicModule: DynamicModule = {
      module: AgvsModule,
    }
    if(options.endpoints && !options.patterns){
      dynamicModule.imports = [
        PassportModule.register({ defaultStrategy: 'jwt' })
      ];
      dynamicModule.controllers = [
        AgvsController
      ]; 
      dynamicModule.providers = [
        {
          provide: 'CONFIG_OPTIONS',
          useValue: options,
        },
        {
          provide: 'SERVICE',
          useClass: AgvsService,
          inject: [AgvsService]
        },
        AgvsService
      ];
      dynamicModule.exports = [
        AgvsService
      ];
    }else if (options.endpoints && options.patterns){
      dynamicModule.imports = [
        PassportModule.register({ defaultStrategy: 'jwt' })
      ];
      dynamicModule.controllers = [
        AgvsController
      ];
      dynamicModule.providers = [
        {
          provide: 'CONFIG_OPTIONS',
          useValue: options,
        },
        {
          provide: 'SERVICE',
          useClass: AgvsPatternsService,
          inject: [AgvsPatternsService]
        },
        AgvsPatternsService
      ];
      dynamicModule.exports = [
        AgvsPatternsService
      ];
    } else if(!options.endpoints && options.patterns) {
      dynamicModule.controllers = [
        AgvsPatternsController
      ]; 
      dynamicModule.providers = [
        {
          provide: 'CONFIG_OPTIONS',
          useValue: options,
        },
        AgvsService,
      ];
      dynamicModule.exports = [
        AgvsService
      ];
    }

    return dynamicModule;
  }
}
