// @moontech-industry/moontech-backend-library/crosses/crosses.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrossesController = void 0;
const common_1 = require("@nestjs/common");
const passport_1 = require("@nestjs/passport");
const swagger_1 = require("@nestjs/swagger");
const crosses_service_1 = require("./crosses.service");
const crosses_dto_1 = require("./dto/crosses.dto");
let CrossesController = class CrossesController {
    constructor(crossesService) {
        this.crossesService = crossesService;
    }
    fetchAllUsers() {
        return this.crossesService.fetchCrosses();
    }
    createCross(newCross) {
        console.log('El cross que envío', newCross);
        return this.crossesService.createCross(newCross);
    }
    updateCross(crossId, newCross) {
        return this.crossesService.updateCross(crossId, newCross);
    }
    deleteCross(crossId) {
        return this.crossesService.deleteCross(crossId);
    }
};
__decorate([
    (0, common_1.Get)(),
    (0, common_1.UseGuards)((0, passport_1.AuthGuard)()),
    (0, swagger_1.ApiBearerAuth)('JWT-auth'),
    (0, swagger_1.ApiOperation)({
        summary: 'Get all crosses. AuthGuard Required',
    }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], CrossesController.prototype, "fetchAllUsers", null);
__decorate([
    (0, common_1.Post)(),
    (0, common_1.UseGuards)((0, passport_1.AuthGuard)()),
    (0, swagger_1.ApiBearerAuth)('JWT-auth'),
    (0, swagger_1.ApiOperation)({
        summary: 'Creates a new cross. AuthGuard Required'
    }),
    __param(0, (0, common_1.Body)(common_1.ValidationPipe)),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [crosses_dto_1.Cross]),
    __metadata("design:returntype", Promise)
], CrossesController.prototype, "createCross", null);
__decorate([
    (0, common_1.Put)('/:crossId'),
    (0, common_1.UseGuards)((0, passport_1.AuthGuard)()),
    (0, swagger_1.ApiBearerAuth)('JWT-auth'),
    (0, swagger_1.ApiOperation)({
        summary: 'Update a cross. AuthGuard Required'
    }),
    __param(0, (0, common_1.Param)('crossId')),
    __param(1, (0, common_1.Body)(common_1.ValidationPipe)),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, crosses_dto_1.Cross]),
    __metadata("design:returntype", Promise)
], CrossesController.prototype, "updateCross", null);
__decorate([
    (0, common_1.Delete)('/:crossId'),
    (0, common_1.UseGuards)((0, passport_1.AuthGuard)()),
    (0, swagger_1.ApiBearerAuth)('JWT-auth'),
    (0, swagger_1.ApiOperation)({
        summary: 'Delete a cross. AuthGuard Required'
    }),
    __param(0, (0, common_1.Param)('crossId')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], CrossesController.prototype, "deleteCross", null);
CrossesController = __decorate([
    (0, common_1.Controller)('api/v1/crosses'),
    (0, swagger_1.ApiTags)('crosses'),
    __metadata("design:paramtypes", [crosses_service_1.CrossesService])
], CrossesController);
exports.CrossesController = CrossesController;
