// @moontech-industry/moontech-backend-library/auth/auth.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthService = void 0;
const common_1 = require("@nestjs/common");
const jwt_1 = require("@nestjs/jwt");
const users_service_1 = require("../users/users.service");
const token_dto_1 = require("./dto/token.dto");
let AuthService = class AuthService {
    constructor(usersService, jwtService) {
        this.usersService = usersService;
        this.jwtService = jwtService;
    }
    async logIn(authCredentialsDto) {
        const userCode = await this.validateUserPassword(authCredentialsDto);
        if (!userCode) {
            throw new common_1.UnauthorizedException('Invalid credentials.');
        }
        const payload = { userCode };
        const accessToken = this.jwtService.sign(payload);
        return new token_dto_1.Token(accessToken, 'jwt', 'WIP');
    }
    async validateUserPassword(authCredentialsDto) {
        const { userCode, password } = authCredentialsDto;
        const user = await this.usersService.findUser(userCode);
        if (user) {
            const validPassword = await user.comparePassword(password);
            if (validPassword) {
                return user.userCode;
            }
            else {
                return null;
            }
        }
        else {
            throw new common_1.UnauthorizedException('Invalid credentials.');
        }
    }
    async fetchUserData(userCode) {
        const user = await this.usersService.fetchUserData(userCode);
        return user;
    }
};
AuthService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [users_service_1.UsersService,
        jwt_1.JwtService])
], AuthService);
exports.AuthService = AuthService;
