// @moontech-industry/moontech-backend-library/plc-opcua-connection/plc-opcua-connection.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var PlcOpcuaConnectionModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlcOpcuaConnectionModule = void 0;
const common_1 = require("@nestjs/common");
const node_operations_module_1 = require("../node-operations/node-operations.module");
const plc_opcua_connection_controller_1 = require("./plc-opcua-connection.controller");
const plc_opcua_connection_service_1 = require("./plc-opcua-connection.service");
let PlcOpcuaConnectionModule = PlcOpcuaConnectionModule_1 = class PlcOpcuaConnectionModule {
    static register(options) {
        return {
            module: PlcOpcuaConnectionModule_1,
            providers: [
                {
                    provide: 'CONFIG_OPTIONS',
                    useValue: options,
                },
                plc_opcua_connection_service_1.PlcOpcuaConnectionService,
            ],
            exports: [plc_opcua_connection_service_1.PlcOpcuaConnectionService]
        };
    }
};
PlcOpcuaConnectionModule = PlcOpcuaConnectionModule_1 = __decorate([
    (0, common_1.Module)({
        imports: [node_operations_module_1.NodeOperationsModule],
        controllers: [plc_opcua_connection_controller_1.PlcOpcuaConnectionController]
    })
], PlcOpcuaConnectionModule);
exports.PlcOpcuaConnectionModule = PlcOpcuaConnectionModule;
