// @moontech-industry/moontech-backend-library/plc-opcua-connection/plc-opcua-connection.constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrafficConnectionError = exports.GLOBAL_ERROR = exports.SESSION_CLOSED = exports.SESSION_SUCCESSFUL = exports.CONECTION_CLOSED = exports.CONECTION_SUCCESSFUL = exports.trafficClient = exports.options = exports.connectionStrategy = void 0;
const node_opcua_client_1 = require("node-opcua-client");
exports.connectionStrategy = {
    initialDelay: 1000,
    maxRetry: 1,
};
exports.options = {
    applicationName: 'trafficPlc',
    connectionStrategy: exports.connectionStrategy,
    securityMode: node_opcua_client_1.MessageSecurityMode.None,
    securityPolicy: node_opcua_client_1.SecurityPolicy.None,
    endpoint_must_exist: false,
};
exports.trafficClient = node_opcua_client_1.OPCUAClient.create(exports.options);
exports.CONECTION_SUCCESSFUL = 'Connection with TRAFIC PLC established';
exports.CONECTION_CLOSED = 'Connection with TRAFIC PLC closed';
exports.SESSION_SUCCESSFUL = 'Session with TRAFIC PLC created';
exports.SESSION_CLOSED = 'Session with TRAFIC PLC closed';
exports.GLOBAL_ERROR = 'An error has occured with AGV PLC : ';
class TrafficConnectionError extends Error {
    constructor(message) {
        super(message);
        this.name = 'TrafficConnectionError';
    }
}
exports.TrafficConnectionError = TrafficConnectionError;
