// @moontech-industry/moontech-backend-library/mongo-db/mongo-db.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var MongoDbModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MongoDbModule = void 0;
const common_1 = require("@nestjs/common");
const mongo_db_service_1 = require("./mongo-db.service");
let MongoDbModule = MongoDbModule_1 = class MongoDbModule {
    static register(options) {
        return {
            module: MongoDbModule_1,
            providers: [
                {
                    provide: 'CONFIG_OPTIONS',
                    useValue: options,
                },
                mongo_db_service_1.MongoDbService,
            ],
            exports: [mongo_db_service_1.MongoDbService],
            global: true
        };
    }
};
MongoDbModule = MongoDbModule_1 = __decorate([
    (0, common_1.Module)({
        controllers: [],
        providers: [mongo_db_service_1.MongoDbService],
        exports: [mongo_db_service_1.MongoDbService],
    })
], MongoDbModule);
exports.MongoDbModule = MongoDbModule;
