// moontech-backend-library/libs/moontech-backend-library/src/users/dto/editUser.dto.ts
import { RoleDto } from '../../roles/dto/role.dto';
import {
  IsArray,
  IsNotEmpty,
  IsOptional,
  IsString,
  Matches,
  MaxLength,
  MinLength,
} from 'class-validator';
export class EditUserDto {
  @IsString()
  @MinLength(4)
  @MaxLength(20)
  userCode: string;

  @IsString()
  @IsNotEmpty()
  @Matches(
    /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
    { message: 'Email format invalid' },
  )
  email: string;

  @IsString()
  @IsNotEmpty()
  role: RoleDto;

  @IsString()
  @IsOptional()
  @MinLength(8)
  @MaxLength(20)
  @Matches(/((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/, {
    message: 'Password too weak.',
  }) // at least one upper case, at least one lower case, at least one number or special character
  password?: string;

  @IsString()
  @IsOptional()
  @MinLength(8)
  @MaxLength(20)
  @Matches(/((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/, {
    message: 'Password too weak.',
  }) // at least one upper case, at least one lower case, at least one number or special character
  newPassword?: string;
}
